### DISCUSSION TITLE
Strange infinite recursion bug (tabs.onActivated, useWebExtensionStorage, browser.tabs.get, browser.storage.local.addListener, watchWithFilter)

### DISCUSSION THREAD

**Discussion by adao** [antfu/vitesse-webext#180] [General]
I'm not sure what is the perpetrator here. I kept running into this recursive error that happened when I was using `useWebExtensionStorage`. I finally "fixed" it in my app by adding the debounceFilter when using `useWebExtensionStorage`. I wasn't satisfied, so I fired up fresh vitesse-webext project to see if I could re-produce and investigate from there. Finally, I got it down to the bare minimum to reproduce the bug.

To reproduce:

Add  to `~/logic/storage`

```javascript
export const objectDemo = useWebExtensionStorage('objectDemo', {})
```

Add these listeners in `~/src/background/main.ts`

```javascript
browser.runtime.onInstalled.addListener((): void => {
  console.log('Extension installed')
  objectDemo.value = {}
})

let previousTabId = 0
browser.tabs.onActivated.addListener(async ({ tabId }) => {
  console.log('tab activated')
  // const tab = await browser.tabs.get(tabId)

  if (previousTabId) {
    objectDemo.value[previousTabId].status = 'inactive'
    objectDemo.value[previousTabId].updatedAt = new Date().toISOString()
  }

  if (objectDemo.value[tabId]) {
    objectDemo.value[tabId].status = 'active'
    objectDemo.value[tabId].updatedAt = new Date().toISOString()
  } else {
    const tab = await browser.tabs.get(tabId)
    objectDemo.value[tabId] = {
      name: tab.title,
      status: 'active',
      updatedAt: new Date().toISOString(),
    }
  }
  previousTabId = tabId
})
```

And in `useWebExtensionStorage.ts`, add a `console.log('storage changes', changes)` to your storage listener function and a `console.log('data changes', data.value)` to your `watchWithFilter` function. 

Now load the extension, open up your service worker console, and navigate between any two tabs. It happens after 3 total tab navigations that I begin seeing an infinite loop of `console.log` statements. Be ready to quickly turn off the extension or else it's going to start eating up your CPU. 

And I've found a *very bizarre* fix for this. If someone could explain it to me, because it's beyond me at this point. In the `browser.tabs.onActivated` listener, if I move the instantiation of tab, `const tab = await browser.tabs.get(tabId)` to the beginning of function instead of in the `else` block, no infinite loop. 

So weird!!! Even though the bug goes away, I don't consider it a fix at all.

I set up the repo [here](https://github.com/adao/bug-reproduction) if anyone wants to pull and explore

**Comment by gxr404** [comment]
- `const tab = await browser.tabs.get(tabId)` moves to the beginning `watch` executed once. [Example](https://play.vuejs.org/#eNp9Ustu2zAQ/JUFL5ERQW6bngzZ6AMp0B7aoM2RF4ZeO3QokuDDdiHo37sk7ThAg9ykmdnZ2V2O7LNz3T4hW7A+SK9chIAxuRU3anDWRxjB46aFg4jyESbYeDvAFVVcccONtCZEsA87WGZZM4oFvJ9mZ0LaZOKJekeoCH+NhE0yMiprcqdvppnByA2AOAhFzTWiK1rItt1e6ISdII8P/0M33EzcXOxKbVQDniw9TeINGDzAHcVWAZvGY7B6jy2RO5RxBstVFUOOc0/FNsWGQi1XJzj7lJqmpAKYWihN8l+elTKU7TSUroVmLaK42OZFWI2dttvKVJD2Uge5viaDFsY1ZV9A9AmzZT+vx6Az0E/EwWkRkf4AxrHse6KuAL1b1cMUx0Um686nqZ/nI5LkIcVI2/kktZJPS87K1jkrJEXJaNHNq5Dwfv6iI2tZDDTFRm27XbCGXkoZjDNpB6c0+l8urz9wRv2rKWdCa3v4UbA8U3vG5SPKp1fwXThmjLM7Wjb6PXL2zEXhtxgrffvnJx7p+5kc7DppUr9B/s7nSzljlX1JZk2xX+hK2u/lvSuzvQ+3x4gmnIfKQcuti54zevxf3xj9Evem+1jq6MJs+gds5yMI)

- Execute twice without moving. [Example](https://play.vuejs.org/#eNp9Ustu2zAQ/JUFL5ERQW6bngzZ6AMp0B7aoM2RF4ZeO3QokuDDdiHo37sk7Tgogtykmdnd2VmO7LNz3T4hW7A+SK9chIAxuRU3anDWRxjB46aFg4jyESbYeDvAFVVcccONtCZEsA87WGZZM4oFvJ9mZ0LaZOKJekeoCH+NhE0yMipr8qRvppnByA3kJt1e6ISdoIoPGRIHociPRnSl/H/RDTcTN5d2RRjVgKeWnjbxBgwe4I5sq4BN4zFYvceWyB3KOIPlqooh27mnYptiQ6aWqxOc+5SaplgAmFooQ/Jf3pU8lHQactdCsxZRXNrmIKzGTtttZSpIudRFrq+pQQvjmrwvIPqEuWU/r8egM9BPxMFpEZH+AMax5D3RVIDerephSsdFJmvm09TP8xFJ8pBipHQ+Sa3k05KzkjpnhSQrGS26eRUS3s9fTGQti4G22KhttwvW0Espi3Em7eCURv/L5fgDZzS/NuVMaG0PPwqWd2rPuHxE+fQKvgvHjHF2R2Gj3yNnz1wUfoux0rd/fuKRvp/Jwa6TJvUb5O98vpQ9VtmXZNZk+4WuuP1e3rsy2/twe4xownmpbLTcuug5o8f/9Y3VL3Zvuo+lji7Mpn9mNSMI)

two write reactive data:

```js
/** 
 if (previousTabId) {
    objectDemo.value[previousTabId].status = 'inactive'
    objectDemo.value[previousTabId].updatedAt = new Date().toISOString()
  }
**/
{
"1982090497":{"name":"page1","status":"inactive","updatedAt":"2024-06-25T14:40:36.452Z"}
}
// await browser.tabs.get(tabId)
/**
 objectDemo.value[tabId] = {
      name: tab.title,
      status: 'active',
      updatedAt: new Date().toISOString(),
    }
**/
{
	"1982090494":{"name":"page2","status":"active","updatedAt":"2024-06-25T14:40:36.452Z"},
	"1982090497":{"name":"page1","status":"inactive","updatedAt":"2024-06-25T14:40:36.452Z"}
}
```


Recursive guessing is because:
```
reactive data change  {a:1} ->  watch ->  setItem localstorage  -> localstaroge change event ->  change reactive vue data -> at this point, reactive data {a:2} triggered the watch again -> infinite.....
reactive data change  {a:2} ->  watch ->  setItem localstorage ->  localstaroge change event  -> .....
```

Try fixing it: 

```ts
  watchWithFilter(
    data,
    async () => {
      console.log('data changes', data.value)
      try {
        if (data.value == null) {
          storageInterface.removeItem(key)
          return
        }
        const currentStorageValue = await storageInterface.getItem(key)
        const setStorageValue = await serializer.write(data.value)

        if (setStorageValue !== currentStorageValue)
          storageInterface.setItem(key, setStorageValue)
      }
      catch (error) {
        onError(error)
      }
    },
    {
      flush,
      deep,
      eventFilter,
    },
  )
```
sorry， Poor English, not very clear expression 😂

**Reply by adao** [reply]
This is great! Thank you for the [play.vuejs]() examples. I definitely have a better idea of what's going on now.

I appreciate the logs you included too. I realized I wasn't even looking at the logs because I kept on closing out the window to save my CPU. I ended up adding a counter to limit the recursive loop so it would eventually stop and I could actually inspect the logs.

So thanks to you pointing out what was actually going on, I was able to get the example in `Popup.vue` working, I just added a Promise like so, 

```javascript
let oldKey
async function updateObject(key) {
  if (oldKey) {
    objectDemo.value[oldKey].updatedAt = new Date().toISOString()
    objectDemo.value[oldKey].status = 'inactive'
  }
  // added this line, and the infinite loop would happen
  await new Promise((resolve) => resolve())
  if (objectDemo.value[key]) {
    objectDemo.value[key].updatedAt = new Date().toISOString()
    objectDemo.value[key].status = 'active'
  } else {
    objectDemo.value[key] = { updatedAt: new Date().toISOString(), name: randomString(10), status: 'active'}
  }
  oldKey = key
}
```

It's pretty much the same thing as the tabs.onActivated listener in the background service worker. One thing I don't really understand is the order of events. On first tab navigation, I get the following:

```
data changes {684856349: {…}}
storage changes {684856349: {…}}
storage changes {684856349: {…}}
read called - key object
data changes {684856349: {…}}
```

The order makes sense here. But I don't know why `data changes` gets printed again at the very end... When the value gets set again in `read`, I would think it's the same value, so the watcher would not get called again. 

Then, the order of events when I click in the popup is like so:

```
storage changes {1: {…}}
storage changes {1: {…}}
read called - key object
data changes {1: {…}}
```

Now, it's missing the first `data changes` event, which seems pretty odd to me as I believe that's what kicks off the chain of events. Again, there's the `data changes` event at the end, and again, I'm not sure how that gets called. 

I'm not at an expert at Javascript, but I'm assuming the awaited Promise in the middle of the code has something to do with Javascript's event loop. Because we're firing off a Promise, execution will continue somewhere else and that's why the changes are flushed sooner and the watcher function executes more than once. If that's the case, then using the debounceFilter to solve the issue actually makes a lot of sense.

Regarding your solution, I wasn't entirely sure *how* it fixed the problem. It definitely worked, but I didn't fully understand how.

Technically, I still can't find the *exact* cause of the bug, because even when I look at the logs, it seems everything happens in the correct order. Even though the chain of events executes twice for one user action, the order of events is still the correct order. Had there been some unintended interleaving of events between the two chain of events, then I would call that out as the obvious cause, but they stayed in their own "chunk", so to speak. Like so, 

```
storage changes {1: {…}}
storage changes {1: {…}}
read called - key object
data changes {1: {…}}
---
storage changes {1: {…}, 2: {…}}
storage changes {1: {…}, 2: {…}}
read called - key object
data changes {1: {…}, 2: {…}}
```

The `---` designates the "chunks" I was referring to. Since it's all in order, I'm still not sure exactly why the loop happens.

Regarding your solution, I decided to poke around to better understand, and found out the following also worked to remove the infinite looping.

```javascript
  watchWithFilter(
    data,
    async () => {
      console.log('data changes', toRaw(data.value))
      try {
        if (data.value == null) {
          storageInterface.removeItem(key)
          return
        }
        // const currentStorageValue = await storageInterface.getItem(key)
        🤣 await new Promise((resolve) => resolve()) 
        const setStorageValue = await serializer.write(data.value)

        // if (setStorageValue !== currentStorageValue)
          storageInterface.setItem(key, setStorageValue)
      }
      catch (error) {
        onError(error)
      }
    },
    {
      flush,
      deep,
      eventFilter,
    },
  )
```

So I think it worked because of some timing mechanism, and not from additional logic.

I'm satisfied with where it's at now. Even though I still don't fully understand the bug, I understand it better than I did before and feel confident the debounceFilter is a reasonable solution. Really appreciate you helping me out with this!

**Reply by gxr404** [reply]
Hi, The previous statement contains an error.

As long as the read function is executed, it will definitely trigger the watch

Because every time the data is changed, a new object is re serialized

```
data.value = await serializer.read(rawValue) as T
```

[Example](https://play.vuejs.org/#eNp9Uk1PGzEU/CtPvpCI1aZVe4o2UT9EpfZQEHD0xTgvYYPXtuznELTa/86zTYADQtrDemb8Zsb2KH563x4SiqXoog69J4hIya+l7QfvAsEIAbcNPCrS9zDBNrgBznjHmbTSamcjgbvbwyrLZqNawtdpfiK0S5ZeqC+MqvhkNWyT1dQ7m53+2NkcRmkhD2kPyiRkfR0jLX/Fd8ZkA7ONIjWH1bpuyBbOYGvcrjIVZMc65/ycBzQwbhD9EigkzMG6Ra3JBXlBOHijCHkFMI6lycSuAJ1f18pl4jKTtc00dYt8PCy5S0Rc44c2vX5YSVH6SFFIjpLRoltUIePd4p2jaARFbrHtd+0+Ost3UIpJod3ge4Ph0udzilKwfx0qhTLGPf4rWO7UnHB9j/rhA3wfjxmT4ipgxHBAKV45UmGHVOmLm/945P9XcnCbZFj9CXmNfAMpZ6yyX8luOPY7XUn7t7yk3u5u48WR0MZTqRw0K6eil4Kf1e9Pqr/F/dZ+L/v4hsX0DBFS7Rs=)

This explains why there will still be a `data changes` call in the end

Why does triggering a watch not result in a dead loop？

```js
watch call ->   setItem localStorage  ->  localStorage change event  ->  change reactive vue data ->  setItem localStorage -> 
At this point, the "localStorage change event " will not trigger again, 
Because the current localStorage is consistent with the localStorage you want to set
```

The next two watch triggers: 

watch call A ->   setItem localStorage  ->  localStorage change event  -> 
    change reactive vue data ->  setItem localStorage -> 
	localStorage change event（At this point, the event was triggered again because the localStorage at this time is B, and the newly set value is A）

watch call B->   setItem localStorage  ->  localStorage change event  ->
  change reactive vue data ->  setItem localStorage -> 
	localStorage change event（At this point, the event was triggered again because the localStorage at this time is A, and the newly set value is B）

...Infinite loop


next
Your warehouse example still has a dead loop when used

Changing it to this may make it more noticeable
```js
   calls++
      if (calls > 40) {
        console.error('loop calls')
        return
      }
```

——————

I have tried various situations and found that


```js
 watchWithFilter(
  ....
        // infinite loop 
         setTimeout(async () => {
           const setStorageValue = await serializer.write(data.value)
           storageInterface.setItem(key, setStorageValue)
         }, 0)

```

```js
 watchWithFilter(
  ....
        // normal
         setTimeout(async () => {
           const setStorageValue = await serializer.write(data.value)
           storageInterface.setItem(key, setStorageValue)
         }, 100)

```



```js
 watchWithFilter(
  ....
       
        const setStorageValue1 = await serializer.write(data.value)
        const currentStorageValue = await storageInterface.getItem(key)
        const setStorageValue2 = await serializer.write(data.value)
		 // infinite loop 
		// storageInterface.setItem(key, setStorageValue1)
        // normal
		storageInterface.setItem(key, setStorageValue2)

```

This means that  Inconsistent serialization between `setStorageValue1` and `setStorageValue2` during two triggers


Why is `await storageInterface.getItem(key)` effective ?

I guess `await storageInterface.getItem(key)` will wait until the last successful setItem before returning


watch call A ->   setItem localStorage  ->  localStorage change event  -> 
    change reactive vue data (Serialize fresh data after waiting for the previous `setItem(B)` (!!  `data.value` is B)) ->  setItem localStorage -> 
	localStorage change event （Not triggered because   localStorage is still B, and the newly set one is also  B）

watch call B->   setItem localStorage (wait previous `setItem(A)` ( fresh `data.value` is B)) ->  localStorage change event  ->
  change reactive vue data -> watch call B -> setItem(B) -> localStorage change event （Not triggered because   localStorage is still B, and the newly set one is also  B）


Of course, my statement may not be correct. Feel free to give me your advice 😂

**Reply by adao** [reply]
Ah! Yes, thank you. Now I understand the root of the recursion. There were a few things wrong with my debugging code 😅

Strange — I thought `await new Promise((resolve) => resolve())` really was fixing it. Maybe it was a fluke yesterday (or I had a bug in my debugging), but yeah, it's not working now. Not to mention, I was console.logging `toRaw(data.value)` which ended up being a reference and so not helpful to look at.

According to my new debug logs, I think watch calls actually go like

watch call A -> setItem localStorage (!! data.value is B) -> localStorage change event ->
change reactive vue data -> setItem localStorage ->
localStorage change event （Not triggered because localStorage is B, and data.value is also B）

watch call B-> setItem localStorage (data.value is B)) -> localStorage change event (Not triggered because localStorage is B, and data.value is also B）

**Reply by adao** [reply]
I don't think `getItem` necessarily waits for the last `setItem`. Or, even if it does, I don't think that's why the code is working here. 

I think it's that the wait is long enough for Javascript to continue executing other code, and when it comes back to execution, `data.value` is already set to B. According to my logs, watch call B doesn't call setItem before watch call A's getItem call.

Here's my output

<img width="730" alt="image" src="https://github.com/antfu-collective/vitesse-webext/assets/906793/fa1d484b-d895-4e6d-b9f5-359d010bd1bd">

call 6 is "A" (executes code up until getItem)
call 7 is "B" (executes code up until getItem)
call 6 executes code from getItem up until setItem (at this point, data.value is what you would expect in B)
call 7 executes code from getItem up until setItem
call 8 (watch call "C" from A setting reactive Vue data in read function)

**Reply by gxr404** [reply]
You're right, it really has nothing to do with `await storageInterface.getItem(key)`, just wait is long enough 

I just don't think the solutions for  `await storageInterface.getItem(key)`, and `debounceFilter` are very suitable.

How long to wait is suitable, depending on the user's environment and whether the computer equipment is relevant？


I checked the code of [vueuse/useStorage](https://github.com/vueuse/vueuse/blob/0c39dd021dacbaaaf3135b0aa171fddb81325880/packages/core/useStorage/index.ts#L269-L282)

I found that the direction was completely wrong 😅


-  Watch `reactive data` is used to change the localstorage
- `localstorage addListener`is used to change `reactive data`

Does the data modified through ` Watch  reactive` not need to be modified through `localstorage addListener` anymore



`localstorage addListener` paused `Watch` listening when triggered, only changing the reactive data

The final solution:

```js
import { StorageSerializers, pausableWatch } from '@vueuse/core'
....

  async function write() {
    try {
      await (data.value == null ? storageInterface.removeItem(key) : storageInterface.setItem(key, await serializer.write(data.value)))
    }
    catch (error) {
      onError(error)
    }
  }

 const { pause: pauseWatch, resume: resumeWatch } = pausableWatch(
    data,
    write,
    { flush, deep, eventFilter },
  )
const listener = async (changes: Record<string, Storage.StorageChange>) => {
      console.log('storage changes', JSON.parse(changes.object.newValue))

      pauseWatch()
      try {
        for (const [key, change] of Object.entries(changes)) {
          await read({
            key,
            newValue: change.newValue as string | null,
          })
        }
      }
      finally {
        resumeWatch()
      }
    }

...
```
I think this may be a more reliable approach. Do you think so 😂

**Reply by adao** [reply]
## 😭 
this looks perfect!

**Comment by adao** [comment]
Success! Thanks @gxr404 !
